### ISSUE TITLE
Creating My Own Topics

### ISSUE THREAD

**Issue by Bam8647** [Monika-After-Story/MonikaModDev#4890]
Okay, so I thought of an issue that could help others know how to practice editing/creating topics on their own, using their personal versions of MAS. I know you guys have a wiki tutorial on how to create topics but I see that it's only for submitting new topics through the GitHub. I'd like to know how to create my own topics with my personal version of MAS. I just need some clarification on some of the topics below:

**1)** I've seen that DDLC mods run only on a specific version of Renpy. What version do I need?  I currently have the latest installment 7.3.0.

**2)** Once I load up renpy, how do I _properly_ select MAS in order to make changes to the files? Last time I attempted to modify the game files, renpy kept on making a new project/game file within the MAS game folder. Where am I messing up? I can never find the modded script files while searching through the game's code.

**3)** Since I'm still fairly new to the coding concept, how might I create a new topic within the other existing dialogues? Just copy and paste each topic's original coding format? Also, If I wanted to create a new branch of topics, would I need to declare a new variable? If so, where and how might I do this?

**4)** Last but most importantly not least, If and when I find myself successful in creating my topics, how might I continue to update the mod without the updates overwriting my changes? Last time I found myself successful in creating my own topics, I found that every time I updated, my topics would get overwritten. Also, do you guys have an updated version of all of Monika's expression codes?

If you could also provide snippets/screenshots, that would also be extremely helpful.

**Thank you!**

**Comment by thezotikrus** [comment]
You don't need RenPy, a good text editor is enough.
To see how topics work look into `script-topics.rpy`. I guess you can make your own `.rpy`, add a topic and then test it (the game will compile in into `.rpyc`). Updates overwrite only mod files, not your custom ones.
[Also you may find this interesting to read.](https://github.com/Monika-After-Story/MonikaModDev/wiki/Contributing-Guidelines)

**Comment by multimokia** [comment]
I should actually rewrite the wiki tutorial, there's a better way to go about it.

Anyway
1. DDLC runs on Ren'Py 6.99.12.4.2187

2. Personally, I don't use the Ren'Py STK to work on my developments for MAS. I simply keep a *second* MAS install using a local persistent (adjused DDLC.py file to point the persistent read/write location to somewhere else) and put `.rpy` files into that to build/test them.

3. To create a topic, you need the full
```renpy
init 5 python:
    addEvent(
        Event(
            persistent.event_database,
            eventlabel="monika_xyz",
            ...
        )
    )

label monika_xyz:
    m "Dialogue"
    return
```

(NOTE: that the database/structure of the event may change depending on where you want it to go)

4. If you want to add personal submodded topic, you'll need to keep them in a `.rpy` file which isn't in the same name as one of the mod files, as whenever you update, that file will be replaced.

**Comment by ThePotatoGuy** [comment]
>version
` 6.99.12.4.2187`

>making changes

I do the following to setup a folder:
1. copy DDLC into a folder
2. copy MAS into the folder (making sure the folder levels match)
3. set renpy sdk's project folder to the folder containing your DDLC+MAS folder

However, the easiest way to add files is to just write rpys and put them in the `game/` folder. RPYs are just text files with a different extension.

>create new topic

See [Dialogue structure](https://github.com/Monika-After-Story/MonikaModDev/wiki/Dialogue-Coding#dialogue-structure) for more info on the specifics of the format.

>new branch of topics

If you mean a new category/grouping, the `category` property in `init 5 block` is used to group topics. There is no set list of these, so adding a new one is simple as just putting what you want as a category in the property. The system will auto gen the grouping.

>updates overwriting changes

AS mentioned above, have a separate rpy file to avoid overwrites.

**Comment by Bam8647** [comment]
Okay so far I have:

**Downloaded Renpy 6.99.12** [x]
**Created a new folder, used to store an original copy of DDLC.**[x]

However, I'm confused as to what you mean by making sure the files for DDLC+MAS match. Do you mean copying and pasting MAS into the DDLC file? For instance: **New File/DDLC/MAS**?

Also, when you say create a new .rpy file, couldn't I copy one of the modded script files and simply rename it along with the extension .rpy? And if I do that, will the game know to search through both the modded and sub-modded script files?

**Edit** Okay so I figured out how to open the modded files. However, whenever I copy and paste one of script files and rename it's extension to rpy, the file opens up empty in renpy. What did I do wrong?

Attached is what my MAS/game folder currently looks like:

![Current MAS MOD Progress](https://user-images.githubusercontent.com/34694679/67151238-c8368100-f288-11e9-9016-b07c0179908e.PNG)


mypersonalscripts.rpy was a copied version of script-topics.rpyc converted to rpy. Did I do this correctly? Because evidently, renpy isn't finding the actual scripts within the rpy.

**Comment by Bam8647** [comment]
This is what my renpy currently looks like whenever I open my "mypersonalscripts.rpy"

**It's Empty. . .**


![A Monika Renpy](https://user-images.githubusercontent.com/34694679/67151333-5cedae80-f28a-11e9-865d-ba945d91aebb.PNG)

**Comment by multimokia** [comment]
That's not one of the files we overwrite. So I don't know what's up with that.

**Comment by Bam8647** [comment]
Yeah I took one of MAS's script files that was suppose to hold all of the topics for me to edit but whenever I converted it to rpy and renamed the file, nothing appeared in renpy which is super weird. I also noticed the file had the same file size as the original, so I have no clue what's going on lol. However, you can see that I'm able to edit the dev files for both expressions and affection checks. I edited some of Monika's dialogue and poses and it worked. @multimokia

**Comment by Bam8647** [comment]
Here's another lovely image of what's happening when I take one of the modded rpy files and open it with renpy.

![AAA](https://user-images.githubusercontent.com/34694679/67153544-dac4b080-f2b0-11e9-9a03-350fea6ab1f4.PNG)

**Comment by multimokia** [comment]
Well, looks like text encoding isn't `utf-8`, but instead set to `binary`

**Comment by Bam8647** [comment]
I figured it out. I took one of the most recent update zip files and just copied, pasted, and renamed some of the script files in my original MAS. Although.. they may not update correctly lol. I found out that having two different script files (for example script-compliments) wouldn't allow renpy to read from two of the same files. So I just deleted the one I already had and converted one of it's matching files from the zip file and just changed the file to rpy. If whatever I said made sense haha!

_Basically, I removed one of the updating mod files lol_ ._. I hope we don't have any new compliments coming soon haha..

**Comment by ThePotatoGuy** [comment]
>copy one of the modded files

the files bundled in release zips are compiled (rpyc), which is not in a text format. its easier to just create a new text document and rename the extension. the renpy engine will auto compile rpy files into rpycs without the sdk present. dont use the same name or else it will overwrite the rpyc and lead to undefined behavior.

**Comment by Bam8647** [comment]
I've been taking some of the rpy files from the latest release zips and copying, pasting, and renaming them into my MAS folder. So they're ready for renpy to read them. But the thing is, renpy won't let me create duplicates of the script files, even if I rename them. The file either goes totally blank or if I get both to show up in renpy, I can only edit the original mod file to make changes to the game. So I just renamed the original mod file.

**TL;DR** _I tried duplicating the rpy files for script topics, while making sure to rename one. But when I opened both the script files up, the changes I made in the renamed, duplicate version didn't show changes to the game when I was editing. Only the original showed changes. So I renamed the original all together lol.._

The point is, If I created a next text document, how would I get the topics I want to add/edit into the mod? _Also, I'm not sure if there is some advance technique used to allow the persistent to read from two files that are essentially the same but with different names._

**Comment by ThePotatoGuy** [comment]
as mentioned, you dont have to make copies of the original rpys, all rpys are standard text files with a different extension. the engine will auto compile rpys on startup and load them in. dont use the same names as the original files unless you are trying to make changes to the originals. you can add topics by following instructions in previous comments about adding topics.

in general:
1. make text file with whatever name u want, with rpy extension like `mytopics.rpy`. 
2. in that file follow instructions to add ur own topic, making sure to copy the required structure defined in the Dialogue Coding wiki page.
3. launch the game with the file in the game/ dir
4. renpy sill compile the code into an rpyc and load it
5. the game will detect the topic and add it appropriately, assuming u have the init 5 block filled out correctly.

**Comment by Bam8647** [comment]
Okay, I'll try this and if I have difficulties I'll upload screen shots of what I'm stuck on. Thank You @ThePotatoGuy

**Comment by Bam8647** [comment]
Attached is an example of me creating a rpy file from a basic text document and copying two different formats to create a topic from the topics script file. The yellow line divides the two different formats I used to create the topics. But the thing is... **it's still not working...**

I also realized on **line 9** there needs to be an **unlocked=true** but whenever I entered it in, python freaked out, saying "true is undefined."

Thoughts?


![AA5](https://user-images.githubusercontent.com/34694679/67165291-faa7b300-f348-11e9-8afc-0a278300b954.PNG)

**Comment by multimokia** [comment]
`true` isn't valid. `True` is. Capitalization is important.

**Comment by Bam8647** [comment]
Below is the code revised, with the correct use of True.

![AA56](https://user-images.githubusercontent.com/34694679/67165541-f3ce6f80-f34b-11e9-936e-525f38a2b645.PNG)

**This is the outcome...**
![AAA45](https://user-images.githubusercontent.com/34694679/67165545-f8932380-f34b-11e9-8637-a5c0272f21c7.PNG)

Do I need to create a new event every time I create a new topic? If so, my god..

**Comment by multimokia** [comment]
There must be a `label` with the same name as the `eventlabel`

**Comment by Bam8647** [comment]
Okay, did that and fixed the error. But the topic is still not showing up when I launch the game. What on earth is going on lol?

Like I said earlier, it seems as if python is just ignoring my sub-modded topics and deciding to only read from the original modded topics.

I can edit the original mod folder and it'll show in the game, but whenever I create a new rpy folder like this, the system only checks for errors without actually implementing it's content into the game.

**Comment by multimokia** [comment]
Probably because it was added to persistent already and needs to be unlocked manually. To change it now, you need to do it after init 7. `mas_getEV('eventlabel_of_your_event').unlocked=True`

**Comment by Bam8647** [comment]
Finally omg, thank you brother @multimokia! 

I'm gonna go cry in a corner and rethink my decision of being a programmer.

In all seriousness though, for future topics is this how I should format them to avoid the mistake I made earlier?

![AA333](https://user-images.githubusercontent.com/34694679/67165814-29c12300-f34f-11e9-9fb8-5c456345be92.PNG)

**Comment by multimokia** [comment]
Well, you don't need to do the init 7 thing every time. Ideally if you get the Event definition correct the first time, it's no issue and you'll never need to do the init 7 block.

**Comment by Bam8647** [comment]
Okay, I'll try and work on avoiding that in the future. But for the most part, I think I got it all down. 

Thanks @ThePotatoGuy  and @multimokia for the help!
